File: backend/app/workers/jobs/analyze.py
"""Feina programada: recurrents, previsio i avisos derivats, espai per espai."""

from __future__ import annotations

import logging

from sqlalchemy import select

from app.db import session_scope
from app.models import Ledger
from app.services.forecast import check_overdrafts
from app.services.recurring import check_missing_occurrences, detect_recurring

logger = logging.getLogger(__name__)


def run_analysis_job() -> str:
    linies: list[str] = []
    with session_scope() as db:
        for ledger in db.scalars(select(Ledger).where(Ledger.is_active.is_(True))):
            stats = detect_recurring(db, ledger.id)
            missing = check_missing_occurrences(db, ledger.id)
            overdrafts = check_overdrafts(db, ledger.id)
            linies.append(
                f"{ledger.name}: {stats}; {missing} rebuts que no han arribat; "
                f"{overdrafts} avisos de descobert"
            )

    resum = " | ".join(linies)
    logger.info("Analisi feta: %s", resum)
    return resum

File: src/workers/jobs/analyze.ts
/**
 * The analysis job: recurring series, missing bills and the overdraft forecast.
 */

import { eq } from "drizzle-orm";

import { db } from "../../db/client.ts";
import { ledgers } from "../../db/schema/index.ts";
import { checkOverdrafts } from "../../services/forecast.ts";
import {
  checkMissingBills,
  detectRecurring,
  summaryRecurring,
} from "../../services/recurring.ts";

export async function analysisJob(): Promise<string> {
  const lines: string[] = [];

  for (const workspace of await db.select().from(ledgers).where(eq(ledgers.isActive, true))) {
    const recurring = await detectRecurring(workspace.id);
    const falten = await checkMissingBills(workspace.id);
    const overdrafts = await checkOverdrafts(workspace);
    lines.push(
      `${workspace.name}: ${summaryRecurring(recurring)}, ${falten} rebuts que falten, ${overdrafts} avisos de descobert`,
    );
  }

  return lines.join("\n") || "no hi ha cap espai actiu";
}
